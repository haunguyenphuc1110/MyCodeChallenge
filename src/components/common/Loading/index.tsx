import React from 'react'
import { View, Text, Modal, ActivityIndicator } from 'react-native'
import Spinner from 'react-native-spinkit'

import { COLORS } from 'Constants'
import styles from './styles'

enum Animation {
  NONE = 'none',
  SLIDE = 'slide',
  FADE = 'fade',
}

enum Size {
  SMALL = 'small',
  LARGE = 'large',
}

interface LoadingProps {
  color?: string
  animation?: Animation
  overlayColor?: string
  size?: Size
  textContent?: string
  textStyle?: object
  visible: boolean
  indicatorStyle?: {}
  customIndicator?: React.ReactNode
  children?: React.ReactNode
  spinnerKey?: string
}

const Loading = ({
  color = COLORS.white,
  animation = Animation.NONE,
  overlayColor = 'rgba(0, 0, 0, 0.6)',
  size = Size.LARGE,
  textContent = '',
  textStyle = {},
  visible = false,
  indicatorStyle = {},
  customIndicator = (
    <Spinner
      size={60}
      type={'ThreeBounce'}
      color={color} />
  ),
  children = null,
  spinnerKey = '',
}: LoadingProps) => {
  const renderDefaultContent = () => {
    return (
      <View style={styles.background}>
        {customIndicator ? (
          customIndicator
        ) : (
          <ActivityIndicator
            color={color}
            size={size}
            style={[styles.activityIndicator, { ...indicatorStyle }]}
          />
        )}
        <View style={[styles.textContainer, { ...indicatorStyle }]}>
          <Text style={[styles.textContent, textStyle]}>{textContent}</Text>
        </View>
      </View>
    )
  }

  const renderSpinner = () => {
    if (!visible) return null

    const spinner = (
      <View
        style={[styles.container, { backgroundColor: overlayColor }]}
        key={spinnerKey ? spinnerKey : `spinner_${Date.now()}`}>
        {children ? children : renderDefaultContent()}
      </View>
    )

    return (
      <Modal
        animationType={animation}
        supportedOrientations={['landscape', 'portrait']}
        transparent
        visible={visible}>
        {spinner}
      </Modal>
    )
  }

  return renderSpinner()
}

export default Loading
