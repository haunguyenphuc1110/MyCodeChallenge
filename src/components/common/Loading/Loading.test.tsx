import React from 'react'
import renderer from 'react-test-renderer'

import Loading from './index'

jest.mock('react-native-spinkit')

describe('Loading', () => {
  describe('snapshot render', () => {
    test('with initial props', () => {
      const tree = renderer.create(<Loading visible={false} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
    test('with visible props is true', () => {
      const tree = renderer.create(<Loading visible={true} customIndicator={null} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
