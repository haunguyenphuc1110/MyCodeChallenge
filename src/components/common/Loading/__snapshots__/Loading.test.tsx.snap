// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading snapshot render with initial props 1`] = `null`;

exports[`Loading snapshot render with visible props is true 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  supportedOrientations={
    Array [
      "landscape",
      "portrait",
    ]
  }
  transparent={true}
  visible={true}
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "bottom": 0,
          "flex": 1,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.6)",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "bottom": 0,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#FFFFFF"
        hidesWhenStopped={true}
        size="large"
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {},
          ]
        }
      />
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "bottom": 0,
              "flex": 1,
              "justifyContent": "center",
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 20,
                "fontWeight": "bold",
                "height": 50,
                "top": 80,
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</Modal>
`;
